aspnetcore/performance/caching/output.md

```diff
@@ -18,44 +18,60 @@ By [Tom Dykstra](https://github.com/tdykstra)
 
 This article explains how to configure output caching middleware in an ASP.NET Core app. For an introduction to output caching, see [Output caching](xref:performance/caching/overview#output-caching).
 
-The output caching middleware can be used in all types of ASP.NET Core apps: Minimal API, Web API with controllers, MVC, and Razor Pages. The sample app is a Minimal API, but every caching feature it illustrates is also supported in the other app types.
+The output caching middleware can be used in all types of ASP.NET Core apps: Minimal API, Web API with controllers, MVC, and Razor Pages. Code examples are provided for both minimal APIs and controller-based APIs, and every caching feature is also supported in the MVC and Razor Pages app types. The controller-based API examples show how to use attributes to configure caching, and the same method of applying attributes to enable and configure caching is used in MVC and Razor Pages apps.
```

Comment:
## Enhancements to Microsoft.AspNetCore.OpenAPI

### Completion enhancements and package install recommendations for OpenAPI package

ASP.NET Core's OpenAPI support ships in an independent package outside of the shared framework. This means that it is difficult for users to independently discover built-in OpenAPI support by leveraging code-completion aids like Intellisense. In .NET 9 Preview 6, we're shipping a completion provider/codefixer combination to help users discover built-in OpenAPI support more easily.

When a user is typing a statement where an OpenAPI-related API is available, the completion provider will provide a recommendation for said API. For example, in the screenshots below, completions for `AddOpenApi` and `MapOpenApi` are provided when a user is entering an invocation statement on support type, such as `IEndpointConventionBuilder`.

<img width="508" alt="image" src="https://github.com/dotnet/AspNetCore.Docs/assets/1857993/1ed9fcb9-7ab7-4239-b7c8-d323724e7222">

When the completion is accepted and the Microsoft.AspNetCore.OpenApi package is not installed, a codefixer will provide a shortcut for automatically installing the dependency in the project.

<img width="586" alt="image" src="https://github.com/dotnet/AspNetCore.Docs/assets/1857993/675baf6f-147f-48a3-bd51-18a0b4eebdf4">

### Support for `[Required]` and `[DefaultValue]` attributes on parameters/properties

When `[Required]` and `[DefaultValue]` attributes are applied on parameters or properties within complex types, the OpenAPI implementation will map these to the `required` and `default` properties in the OpenAPI document associated with the parameter or type schema.

For example, the following API will produce the accompanying schema for the `Todo` type.

```csharp
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;

var builder = WebApplication.CreateBuilder();

builder.Services.AddOpenApi();

var app = builder.Build();

if (app.Environment.IsDevelopment())
{
    app.MapOpenApi();
}

app.MapPost("/todos", (Todo todo) => { });

app.Run();

class Todo
{
	public int Id { get; init; }
	public required string Title { get; init; }
	[DefaultValue("A new todo")]
	public required string Description { get; init; }
	[Required]
	public DateTime CreatedOn { get; init; }
}
```

```json
{
	"required": [
	  "title",
	  "description",
	  "createdOn"
	],
	"type": "object",
	"properties": {
	  "id": {
	    "type": "integer",
	    "format": "int32"
	  },
	  "title": {
	    "type": "string"
	  },
	  "description": {
	    "type": "string",
	    "default": "A new todo"
	  },
	  "createdOn": {
	    "type": "string",
	    "format": "date-time"
	  }
	}
}
```

### Support for schema transformers on OpenAPI document

In .NET 9 Preview 6, built-in OpenAPI support ships with support for schema transformers that can be used to modify schemas generated by System.Text.Json and the OpenAPI implementation. Like document and operation transformers, schema transformers can be registered on the `OpenApiOptions` object. For example, the code sample below demonstrates using a schema transformer to add an example to type's schema.

```csharp
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using Microsoft.OpenApi.Any;

var builder = WebApplication.CreateBuilder();

builder.Services.AddOpenApi(options =>
{
    options.UseSchemaTransformer((schema, context, cancellationToken) =>
    {
        if (context.Type == typeof(Todo))
        {
            schema.Example = new OpenApiObject
            {
                ["id"] = new OpenApiInteger(1),
                ["title"] = new OpenApiString("A short title"),
                ["description"] = new OpenApiString("A long description"),
                ["createdOn"] = new OpenApiDateTime(DateTime.Now)
            };
        }
        return Task.CompletedTask;
    });
});

var app = builder.Build();

if (app.Environment.IsDevelopment())
{
    app.MapOpenApi();
}

app.MapPost("/todos", (Todo todo) => { });

app.Run();

class Todo
{
	public int Id { get; init; }
	public required string Title { get; init; }
	[DefaultValue("A new todo")]
	public required string Description { get; init; }
	[Required]
	public DateTime CreatedOn { get; init; }
}
```



```suggestion
The output caching middleware can be used in all types of ASP.NET Core apps: Minimal API, Web API with controllers, MVC, and Razor Pages. Code examples are provided for both minimal APIs and controller-based APIs. Every caching feature is also supported in the MVC and Razor Pages. The controller-based API examples show how to use attributes to configure caching. The attributes to enable and configure caching is available to MVC and Razor Pages apps.
```

Reply:
> Every caching feature is also supported in MVC and Razor Pages.

Isn't it already mentioned in the first sentence of this paragraph?: "... with controllers, MVC, and Razor Pages"